src/core/diffusion_grid.h

```diff
@@ -559,6 +565,204 @@ class DiffusionGrid {
     c1_.swap(c2_);
   }
 
+  void RK() {
+    // check if diffusion coefficient and decay constant are 0
+    // i.e. if we don't need to calculate diffusion update
+    if (IsFixedSubstance()) {
+      return;
+    }
+
+    r1_.resize(total_num_boxes_);
+    r2_.resize(total_num_boxes_);
+    r3_.resize(total_num_boxes_);
```

Comment:
You shouldn't resize here. Resizing should only be done in the Update() call. You would otherwise be unnecessarily resizing if the total number of boxes remains the same over iterations.

Just have a check inside the Update() function if the user selected the RK method, because for the other methods we don't need the r containers.